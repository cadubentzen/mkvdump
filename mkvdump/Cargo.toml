[package]
name = "mkvdump"
version = "0.5.0"
authors = ["Carlos Bentzen <cadubentzen@gmail.com>"]
edition = "2021"
license = "MIT OR Apache-2.0"
description = "MKV and WebM parser CLI tool"
repository = "https://github.com/cadubentzen/mkvdump"
keywords = ["mkv", "matroska", "webm", "cli", "ebml"]
categories = ["command-line-utilities", "multimedia", "parser-implementations"]
exclude = ["tests/*", ".github/*"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
mkvparser = { path = "../mkvparser", version = "0.4.4" }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
serde_yaml = "0.9"
clap = { version = "4", features = ["derive"] }

[dev-dependencies]
assert_cmd = "2"
insta = { version = "1.18.1", features = ["yaml"] }
