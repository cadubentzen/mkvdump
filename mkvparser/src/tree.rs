///! The tree module contains helpers for building tree
///! structures from parsed elements
use serde::Serialize;

use crate::{Body, Element, Header, Id};

/// A Master Element that owns its children for diplaying
/// it in an element tree
#[derive(Debug, PartialEq, Serialize)]
pub struct MasterElement {
    #[serde(flatten)]
    header: Header,
    children: Vec<ElementTree>,
}

/// An Element Tree can either be a leaf or a Master
/// element.
#[derive(Debug, PartialEq, Serialize)]
#[serde(untagged)]
pub enum ElementTree {
    /// A Normal Element that represents a leaf in the tree
    Normal(Element),
    /// A Master Element contains more elements.
    Master(MasterElement),
}

impl Id {
    fn can_be_children_of(&self, other: &Id) -> bool {
        !matches!((self, other), (Id::Cluster, Id::Cluster) | (Id::Ebml, _))
    }
}

/// Build element trees from a series of elements
pub fn build_element_trees(elements: &[Element]) -> Vec<ElementTree> {
    let mut trees = Vec::<ElementTree>::new();

    let mut index = 0;
    while index < elements.len() {
        let element = &elements[index];
        match element.body {
            Body::Master => {
                // parse_header() already handles Unknown sizes.
                let mut size_remaining = element.header.body_size.unwrap_or(usize::MAX);

                let mut children = Vec::<Element>::new();
                while size_remaining > 0 {
                    index += 1;

                    if let Some(next_child) = elements.get(index) {
                        if !next_child.header.id.can_be_children_of(&element.header.id) {
                            index -= 1;
                            break;
                        }

                        size_remaining -= if let Body::Master = next_child.body {
                            // Master elements' body size should not count in the recursion
                            // as the children would duplicate the size count, so
                            // we only consider the header size on the calculation.
                            next_child.header.header_size
                        } else {
                            next_child
                                .header
                                .size
                                .expect("Only Master elements can have unknown size")
                        };
                        children.push(next_child.clone());
                    } else {
                        // Elements have ended before reaching the size of the master element
                        break;
                    }
                }
                trees.push(ElementTree::Master(MasterElement {
                    header: element.header.clone(),
                    children: build_element_trees(&children),
                }));
            }
            _ => {
                trees.push(ElementTree::Normal(element.clone()));
            }
        }
        index += 1;
    }
    trees
}

#[cfg(test)]
mod tests {
    use crate::Unsigned;

    use super::*;

    #[test]
    fn test_build_element_trees() {
        let elements = [
            Element {
                header: Header::new(Id::Ebml, 5, 31),
                body: Body::Master,
            },
            Element {
                header: Header::new(Id::EbmlVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(1)),
            },
            Element {
                header: Header::new(Id::EbmlReadVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(1)),
            },
            Element {
                header: Header::new(Id::EbmlMaxIdLength, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(4)),
            },
            Element {
                header: Header::new(Id::EbmlMaxSizeLength, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(8)),
            },
            Element {
                header: Header::new(Id::DocType, 3, 4),
                body: Body::String("webm".to_string()),
            },
            Element {
                header: Header::new(Id::DocTypeVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(4)),
            },
            Element {
                header: Header::new(Id::DocTypeReadVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(2)),
            },
        ];

        let expected = vec![ElementTree::Master(MasterElement {
            header: Header::new(Id::Ebml, 5, 31),
            children: vec![
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(1)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlReadVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(1)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlMaxIdLength, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(4)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlMaxSizeLength, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(8)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocType, 3, 4),
                    body: Body::String("webm".to_string()),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocTypeVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(4)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocTypeReadVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(2)),
                }),
            ],
        })];

        assert_eq!(build_element_trees(&elements), expected);
    }
}
