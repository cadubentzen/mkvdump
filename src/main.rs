#![doc = include_str!("../README.md")]

use clap::{Parser, ValueEnum};
use mkvdump::parse_elements_from_file;
use mkvparser::tree::build_element_trees;
use serde::Serialize;

#[doc(hidden)]
#[derive(Parser)]
#[clap(author, version, about, long_about = None)]
struct Args {
    /// Name of the MKV/WebM file to be parsed
    filename: String,

    /// Output format
    #[clap(value_enum, short, long, default_value = "yaml")]
    format: Format,

    /// Add element positions in the output
    #[clap(short = 'p', long)]
    show_element_positions: bool,

    /// Show output as a sequence, rather than a tree
    #[clap(short = 'l', long)]
    linear_output: bool,
}

#[doc(hidden)]
#[derive(ValueEnum, Clone, PartialEq, Eq)]
enum Format {
    Json,
    Yaml,
}

#[doc(hidden)]
fn print_serialized<T: Serialize>(elements: &[T], format: &Format) {
    let serialized = match format {
        Format::Json => serde_json::to_string_pretty(elements).unwrap(),
        Format::Yaml => serde_yaml::to_string(elements).unwrap(),
    };
    println!("{}", serialized);
}

#[doc(hidden)]
fn main() -> anyhow::Result<()> {
    let args = Args::parse();
    let elements = parse_elements_from_file(&args.filename, args.show_element_positions)?;

    if args.linear_output {
        print_serialized(&elements, &args.format);
    } else {
        let element_trees = build_element_trees(&elements);
        print_serialized(&element_trees, &args.format);
    }

    Ok(())
}
