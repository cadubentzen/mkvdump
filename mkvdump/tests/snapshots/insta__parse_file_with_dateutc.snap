---
source: src/main.rs
assertion_line: 898
expression: "parse_buffer_to_end(include_bytes!(\"../inputs/dateutc.mkv\"))"
---
- id: EBML
  header_size: 5
  size: 40
  children:
    - id: EBMLVersion
      header_size: 3
      size: 4
      value: 1
    - id: EBMLReadVersion
      header_size: 3
      size: 4
      value: 1
    - id: EBMLMaxIDLength
      header_size: 3
      size: 4
      value: 4
    - id: EBMLMaxSizeLength
      header_size: 3
      size: 4
      value: 8
    - id: DocType
      header_size: 3
      size: 11
      value: matroska
    - id: DocTypeVersion
      header_size: 3
      size: 4
      value: 4
    - id: DocTypeReadVersion
      header_size: 3
      size: 4
      value: 2
- id: Segment
  header_size: 12
  size: 3859
  children:
    - id: SeekHead
      header_size: 5
      size: 70
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[c4 b0 cf 83]"
        - id: Seek
          header_size: 3
          size: 14
          children:
            - id: SeekID
              header_size: 3
              size: 7
              value: Info
            - id: SeekPosition
              header_size: 3
              size: 4
              value: 229
        - id: Seek
          header_size: 3
          size: 15
          children:
            - id: SeekID
              header_size: 3
              size: 7
              value: Tracks
            - id: SeekPosition
              header_size: 3
              size: 5
              value: 320
        - id: Seek
          header_size: 3
          size: 15
          children:
            - id: SeekID
              header_size: 3
              size: 7
              value: Tags
            - id: SeekPosition
              header_size: 3
              size: 5
              value: 445
        - id: Seek
          header_size: 3
          size: 15
          children:
            - id: SeekID
              header_size: 3
              size: 7
              value: Cues
            - id: SeekPosition
              header_size: 3
              size: 5
              value: 3819
    - id: Void
      header_size: 9
      size: 159
      value: ~
    - id: Info
      header_size: 5
      size: 91
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[c9 1e b8 5d]"
        - id: TimestampScale
          header_size: 4
          size: 7
          value: 1000000
        - id: MuxingApp
          header_size: 3
          size: 16
          value: Lavf58.29.100
        - id: WritingApp
          header_size: 3
          size: 16
          value: Lavf58.29.100
        - id: SegmentUUID
          header_size: 3
          size: 19
          value: "[93 f3 31 4f b3 5e e7 a1 a4 db f2 52 90 f7 7c 2d]"
        - id: DateUTC
          header_size: 3
          size: 11
          value: "2022-08-11T08:27:15Z"
        - id: Duration
          header_size: 3
          size: 11
          value: 80
    - id: Tracks
      header_size: 5
      size: 125
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[e4 1f 42 9b]"
        - id: TrackEntry
          header_size: 9
          size: 114
          children:
            - id: TrackNumber
              header_size: 2
              size: 3
              value: 1
            - id: TrackUID
              header_size: 3
              size: 4
              value: 1
            - id: FlagLacing
              header_size: 2
              size: 3
              value: 0
            - id: Language
              header_size: 4
              size: 7
              value: und
            - id: CodecID
              header_size: 2
              size: 17
              value: V_MPEG4/ISO/AVC
            - id: TrackType
              header_size: 2
              size: 3
              value: video
            - id: DefaultDuration
              header_size: 4
              size: 8
              value: 40000000
            - id: Video
              header_size: 9
              size: 16
              children:
                - id: PixelWidth
                  header_size: 2
                  size: 4
                  value: 320
                - id: PixelHeight
                  header_size: 2
                  size: 3
                  value: 240
            - id: CodecPrivate
              header_size: 3
              size: 44
              value: "[01 f4 00 0d ff e1 00 19 67 f4 00 0d 91 9b 28 28 3f 60 22 00 00 03 00 02 00 00 03 00 64 1e 28 53 2c 01 00 05 68 eb e3 c4 48]"
    - id: Tags
      header_size: 6
      size: 204
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[85 0b 7d 15]"
        - id: Tag
          header_size: 10
          size: 56
          children:
            - id: Targets
              header_size: 10
              size: 10
              children: []
            - id: SimpleTag
              header_size: 10
              size: 36
              children:
                - id: TagName
                  header_size: 3
                  size: 10
                  value: ENCODER
                - id: TagString
                  header_size: 3
                  size: 16
                  value: Lavf58.29.100
        - id: Tag
          header_size: 10
          size: 68
          children:
            - id: Targets
              header_size: 10
              size: 14
              children:
                - id: TagTrackUID
                  header_size: 3
                  size: 4
                  value: 1
            - id: SimpleTag
              header_size: 10
              size: 44
              children:
                - id: TagName
                  header_size: 3
                  size: 10
                  value: ENCODER
                - id: TagString
                  header_size: 3
                  size: 24
                  value: Lavc58.54.100 libx264
        - id: Tag
          header_size: 10
          size: 68
          children:
            - id: Targets
              header_size: 10
              size: 14
              children:
                - id: TagTrackUID
                  header_size: 3
                  size: 4
                  value: 1
            - id: SimpleTag
              header_size: 10
              size: 44
              children:
                - id: TagName
                  header_size: 3
                  size: 11
                  value: DURATION
                - id: TagString
                  header_size: 3
                  size: 23
                  value: "00:00:00.080000000"
    - id: Cluster
      header_size: 6
      size: 3170
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[81 99 ab 1f]"
        - id: Timestamp
          header_size: 2
          size: 3
          value: 0
        - id: SimpleBlock
          header_size: 3
          size: 2857
          value:
            track_number: 1
            timestamp: 0
            keyframe: true
        - id: SimpleBlock
          header_size: 3
          size: 298
          value:
            track_number: 1
            timestamp: 40
    - id: Cues
      header_size: 5
      size: 28
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[92 28 61 ef]"
        - id: CuePoint
          header_size: 2
          size: 17
          children:
            - id: CueTime
              header_size: 2
              size: 3
              value: 0
            - id: CueTrackPositions
              header_size: 2
              size: 12
              children:
                - id: CueTrack
                  header_size: 2
                  size: 3
                  value: 1
                - id: CueClusterPosition
                  header_size: 2
                  size: 4
                  value: 649
                - id: CueRelativePosition
                  header_size: 2
                  size: 3
                  value: 9

